File: 백준/레벨Gold IV/1043. 거짓말/거짓말.py
from sys import stdin

input = stdin.readline

def disjoint_set():
    # 부모노드 초기화
    N, M = map(int, input().split())
    parent = list(range(N+1))

    def find(x):
        nodes = []
        while parent[x] != x:
            nodes.append(x)
            x = parent[x]

        for node in nodes:
            parent[node] = x
        return x
    
    def union(a, b):
        root_a = find(a)
        root_b = find(b)
        
        if root_a > root_b:
            parent[root_a] = root_b
        else:
            parent[root_b] = root_a

    # 진실노드 초기화
    truth = list(map(int, input().split())) 
    if truth[0] == 0:
        print(M)
        return

    # 파티노드 초기화
    parties = []
    for _ in range(M):
        party = list(map(int, input().split()))[1:]
        parties.append(party)
        
        for i in range(len(party) - 1):
            union(party[i], party[i + 1])
    
    # 진실노드로 병합
    root = find(truth[1])
    for i in range(2, len(truth)):
        union(truth[i], root)

    # 파티 처리
    count = M
    for party in parties:
        if find(party[0]) == find(root):
            count -= 1
    
    print(count)

disjoint_set()

File: 백준/레벨Gold IV/1043. 거짓말/거짓말.ts
// @ts-ignore
import fs = require('fs');

const read = fs.readFileSync(0, 'utf-8');
const lines = read.split('\n').map((line: string) => line.split(' ').map(Number));

const disjointSet = () => {
  // 부모노드 초기화
  const [N, M] = lines[0];
  const parent = Array.from({ length: N + 1 }, (_, i) => i);

  const find = (x: number): number => {
    const nodes: number[] = [];
    while (parent[x] !== x) {
      nodes.push(x);
      x = parent[x];
    }

    for (const node of nodes) parent[node] = x;
    return x;
  };

  const union = (a: number, b: number): void => {
    const root_a = find(a);
    const root_b = find(b);

    if (root_a > root_b) parent[root_a] = root_b;
    else parent[root_b] = root_a;
  };

  // 진실노드 초기화
  const truth = lines[1];
  if (truth[0] === 0) {
    console.log(M);
    return;
  }

  // 파티노드 초기화
  const parties: number[][] = [];
  for (let i = 2; i <= M + 1; i++) {
    const party = lines[i].slice(1);
    parties.push(party);

    for (let j = 0; j < party.length - 1; j++) {
      union(party[j], party[j + 1]);
    }
  }

  // 진실노드로 병합
  const root = find(truth[1]);
  for (let i = 2; i < truth.length; i++) {
    union(truth[i], root);
  }

  // 파티 처리
  let count = M;
  for (const party of parties) {
    if (find(party[0]) === find(root)) count--;
  }

  console.log(count);
};

disjointSet();